### PR TITLE
config: Azure SNP tool can delete specific version from attestation API

### PR DESCRIPTION
**Pull request by elchead**
### Proposed change(s)

- add unit tests for attestationconfig.Client
- add delete command to tool
- refactor attestationconfig.Client to use the general apiclient

TODO: also make rest of attestation client testable?

### Checklist

- [ ] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [x] Add labels (e.g., for changelog category)
- [ ] Link to Milestone

### CODE REVIEW

File: hack/configapi/cmd/delete.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+)
+
+// newDeleteCmd creates the delete command.
+func newDeleteCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "delete",
+		Short: "delete a specific version from the config api",
+		RunE:  runDelete,
+	}
+	cmd.PersistentFlags().StringP("version", "v", "", "Name of the version to delete (without .json suffix)")
```
Comment (daniel-weisse) [inline hack/configapi/cmd/delete.go:24]:
```suggestion
	cmd.Flags().StringP("version", "v", "", "Name of the version to delete (without .json suffix)")
```
nit: If the flag is just used for this command, and not for any sub commands of `delete`, just use the regular `Flags`

File: hack/configapi/cmd/delete.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	defer func() {
+		if d.closefn != nil {
+			if err := d.closefn(ctx); err != nil {
+				fmt.Printf("close client: %s\n", err.Error())
+			}
+		}
+	}()
+	version := cmd.Flag("version").Value.String()
+	return d.attestationclient.DeleteAzureSEVSNVersion(ctx, version)
+}
```
Comment (daniel-weisse) [inline hack/configapi/cmd/delete.go:44]:
```suggestion
func (d deleteCmd) delete(cmd *cobra.Command) error {
	defer func() {
		if d.closefn != nil {
			if err := d.closefn(ctx); err != nil {
				cmd.Printf("close client: %s\n", err.Error())
			}
		}
	}()
	version := cmd.Flag("version").Value.String()
	return d.attestationclient.DeleteAzureSEVSNVersion(cmd.Context(), version)
}
```
* Don't need to pass an explicit context if the function already uses `*cobra.Command`, as the command already has a context embedded
* Use `cmd.PrintX` instead of `fmt.PrintX`

File: hack/configapi/cmd/delete.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+func runDelete(cmd *cobra.Command, _ []string) error {
+	ctx := context.Background()
+	cfg := staticupload.Config{
+		Bucket: awsBucket,
+		Region: awsRegion,
+	}
+	repo, closefn, err := client.New(ctx, cfg, []byte(cosignPwd), []byte(privateKey))
```
Comment (daniel-weisse) [inline hack/configapi/cmd/delete.go:52]:
```suggestion
	cfg := staticupload.Config{
		Bucket: awsBucket,
		Region: awsRegion,
	}
	repo, closefn, err := client.New(cmd.Context(), cfg, []byte(cosignPwd), []byte(privateKey))
```
Same here

File: hack/configapi/cmd/delete.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+		RunE:  runDelete,
+	}
+	cmd.PersistentFlags().StringP("version", "v", "", "Name of the version to delete (without .json suffix)")
+	must(enforceRequiredFlags(cmd, "version"))
+	return cmd
+}
+
+type deleteCmd struct {
+	attestationclient client.Client
+	closefn           client.CloseFunc
```
Comment (daniel-weisse) [inline hack/configapi/cmd/delete.go:31]:
```suggestion
	attestationClient client.Client
	closeFn           client.CloseFunc
```
nit: naming

File: hack/configapi/cmd/delete.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+		Short: "delete a specific version from the config api",
+		RunE:  runDelete,
+	}
+	cmd.PersistentFlags().StringP("version", "v", "", "Name of the version to delete (without .json suffix)")
+	must(enforceRequiredFlags(cmd, "version"))
+	return cmd
+}
+
+type deleteCmd struct {
+	attestationclient client.Client
```
Comment (daniel-weisse) [inline hack/configapi/cmd/delete.go:30]: Instead of importing the `client.Client` interface here, define a new interface in this file that only has the relevant functions for `delete`.

Example:
```Go
type deleteClient interface {
    DeleteAzureSEVSNVersion(ctx context.Context, versionStr string) error
}

That way the implementation of delete is a lot more flexible, since we only depend on the client implementation to implement the function we actually use in the code, and writing unit test is a bit easier, since we can skip out on writing stub functions that are never called.

File: hack/configapi/cmd/delete_test.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+	return nil
+}
+
+func (f fakeAttestationClient) DeleteList(_ context.Context, _ variant.Variant) error {
+	return nil
+}
+
+func (f fakeAttestationClient) List(_ context.Context, _ variant.Variant) ([]string, error) {
+	return []string{}, nil
+}
```
Comment (daniel-weisse) [inline hack/configapi/cmd/delete_test.go:53]:
```suggestion
```

Can be removed if  the interface used for `attestationclient` is more restricted, see my previous comment

File: internal/api/attestationconfig/client/client.go [hunk 7]
```diff
@@ -19,51 +19,60 @@ import (
...
 
-// Client manages (modifies) the version information for the attestation variants.
-type Client struct {
+// Client is the client for the attestationconfig API.
+type Client interface {
+	UploadAzureSEVSNP(ctx context.Context, versions attestationconfig.AzureSEVSNPVersion, date time.Time) error
+	List(ctx context.Context, attestation variant.Variant) ([]string, error)
+	DeleteList(ctx context.Context, attestation variant.Variant) error
+	DeleteAzureSEVSNVersion(ctx context.Context, versionStr string) error
+}
```
Comment (daniel-weisse) [inline internal/api/attestationconfig/client/client.go:36]: Unless there are a lot of places using this interface, prefer defining the interface at the place it is used, instead of the place where the implementation is.
See my previous comment for the `delete` command.

File: internal/api/attestationconfig/client/client.go [hunk 8]
```diff
@@ -121,7 +130,56 @@ func (a Client) DeleteList(ctx context.Context, attestation variant.Variant) err
...
+	versionStr = versionStr + ".json"
+	variant := variant.AzureSEVSNP{}
+	// delete version file
+	filePath := fmt.Sprintf("%s/%s/%s", constants.CDNAttestationConfigPrefixV1, variant.String(), versionStr)
+	ops = append(ops, deleteCmd{
+		path: filePath,
+	})
+
+	// delete signature file
+	sigPath := fmt.Sprintf("%s/%s/%s", constants.CDNAttestationConfigPrefixV1, variant.String(), versionStr+".sig")
```
Comment (daniel-weisse) [inline internal/api/attestationconfig/client/client.go:143]:
```suggestion
	filePath := path..Join(constants.CDNAttestationConfigPrefixV1, variant.String(), versionStr)
	ops = append(ops, deleteCmd{
		path: filePath,
	})

	// delete signature file
	sigPath :=path..Join(constants.CDNAttestationConfigPrefixV1, variant.String(), versionStr+".sig")
```
I don't have a strong preference of using `fmt.Sprintf` over `path.Join` (or the other way around), but we should try to keep the functions used to generate the url paths consistent, e.g. [line 159](https://github.com/edgelesssys/constellation/blob/ed8fb1c89cd9f13db110e009e1fbde3d88409378/internal/api/attestationconfig/client/client.go#L159) uses `path.Join` instead.

File: internal/api/attestationconfig/client/client.go [hunk 9]
```diff
@@ -166,14 +224,58 @@ func put(ctx context.Context, client s3Client, bucket, path string, data []byte)
...
+	) (*s3.DeleteObjectOutput, error)
 }
 
 // CloseFunc is a function that closes the client.
 type CloseFunc func(ctx context.Context) error
+
+func removeVersion(versions attestationconfig.AzureSEVSNPVersionList, versionStr string) (removedVersions attestationconfig.AzureSEVSNPVersionList, err error) {
+	for i, v := range versions {
+		if v == versionStr {
+			removedVersions = append(versions[:i], versions[i+1:]...)
```
Comment (daniel-weisse) [inline internal/api/attestationconfig/client/client.go:255]: Need to make sure `versions[i+1]` actually exists

File: internal/api/attestationconfig/client/client_test.go [hunk 10]
```diff
@@ -1,82 +1,51 @@
...
-	flag.Parse()
-	if *cosignPwd == "" || *privateKeyPath == "" {
-		flag.Usage()
-		fmt.Println("Required flags not set: --cosign-pwd, --private-key. Skipping tests.")
-		os.Exit(1)
-	}
-	if _, present := os.LookupEnv(envAwsKey); !present {
-		fmt.Printf("%s not set. Skipping tests.\n", envAwsKey)
-		os.Exit(1)
+// TODO replace with unit testable version
```
Comment (daniel-weisse) [inline internal/api/attestationconfig/client/client_test.go:17]: Fix TODO or add ticket reference and/or add person responsible to implement this

**Review by daniel-weisse — COMMENTED** [review]
Just some style things

File: internal/api/attestationconfig/client/client_test.go (hunk 10 above)
Comment (elchead) [reply internal/api/attestationconfig/client/client_test.go:17]: i wanted to get feedback on this way of testing. You think we should use it in all parts?

File: internal/api/attestationconfig/azure.go [hunk 11]
```diff
@@ -34,14 +34,14 @@ type AzureSEVSNPVersion struct {
 	Microcode uint8 `json:"microcode"`
 }
 
-// AzureSEVSNPVersionGet is the request to get the version information of the specific version in the config api.
-type AzureSEVSNPVersionGet struct {
+// AzureSEVSNPVersionAPI is the request to get the version information of the specific version in the config api.
+type AzureSEVSNPVersionAPI struct {
```
Comment (derpsteb) [inline internal/api/attestationconfig/azure.go:38]: Why did you rename this? And what is the rationale behind the name? 
The previous name seemed pretty intuitive to me and follows the pattern of other objects within the Resource API.

Comment (elchead) [reply internal/api/attestationconfig/azure.go:38]: The Resource can be used for upload and get. Do u have a better naming suggestion?

Comment (derpsteb) [reply internal/api/attestationconfig/azure.go:38]: Makes sense :+1: Maybe adapt the comment?
I guess it should just be named `AzureSEVSNPVersion`. But we already have that. I guess we can fix that when refactoring this to use `AttestationCfg`.

File: hack/configapi/cmd/root.go [hunk 12]
```diff
@@ -94,7 +96,7 @@ func runCmd(cmd *cobra.Command, _ []string) error {
 	}
 	if isNewer {
 		fmt.Printf("Input version: %+v is newer than latest API version: %+v\n", inputVersion, latestAPIVersion)
```
Comment (daniel-weisse) [inline hack/configapi/cmd/root.go:98]:
```suggestion
		cmd.Printf("Input version: %+v is newer than latest API version: %+v\n", inputVersion, latestAPIVersion)
```

File: internal/api/attestationconfig/client/client.go [hunk 13]
```diff
@@ -9,49 +9,49 @@ import (
...
 		cosignPwd:     cosignPwd,
 		privKey:       privateKey,
+		fetcher:       fetcher.New(),
 	}
-	repoClose := func(ctx context.Context) error {
-		return repo.Close(ctx)
-	}
-	return repo, repoClose, nil
+	return repo, repo.Close, nil
 }
```
Comment (daniel-weisse) [inline internal/api/attestationconfig/client/client.go:55]: Why not just return the close function returned by `apiclient.NewClient`?

File: internal/api/attestationconfig/client/client.go [hunk 14]
```diff
@@ -99,60 +99,142 @@ func (a Client) createAndUploadSignature(ctx context.Context, content []byte, fi
...
+	Upload(
+		ctx context.Context, input *s3.PutObjectInput, opts ...func(*s3manager.Uploader),
+	) (*s3manager.UploadOutput, error)
+	DeleteObject(ctx context.Context, params *s3.DeleteObjectInput,
+		optFns ...func(*s3.Options),
+	) (*s3.DeleteObjectOutput, error)
+}
+
+// CloseFunc is a function that closes the client.
+type CloseFunc func(ctx context.Context) error
```
Comment (daniel-weisse) [inline internal/api/attestationconfig/client/client.go:193]: Doesnt need to be exported (or defined at all probably).

File: internal/api/attestationconfig/client/client_test.go (hunk 10 above)
Comment (daniel-weisse) [reply internal/api/attestationconfig/client/client_test.go:17]: I prefer having unit tests over integration tests that need to be ran manually in packages.
That being said, it might still be helpful to have an integration test to test the functionality of a package against a real API.
If you want to keep a test like that, I suggest moving it into a separate directory and package, e.g. `internal/api/attestationconfig/client/test/integration_test.go

Comment (derpsteb) [reply internal/api/attestationconfig/client/client_test.go:17]: I like the idea for testing, but would not extend it to the other tests within this PR.

**Review by derpsteb — APPROVED** [review]
